File: backend/app/services/sku_costs.py
"""
sku_costs.py — SKU cost file management.

Handles both v1 flat format ({"SKU": 60}) and v2 split format
({"SKU": {"making_cost": 60, "packaging_cost": 8}}).  The API always
exposes the v2 schema; v1 entries are transparently upgraded on read
(total_cost → making_cost, packaging_cost = 0).
"""

from __future__ import annotations

import json
from pathlib import Path


def load_costs(path: Path) -> dict[str, dict]:
    """Read the SKU cost JSON and normalise every entry to v2 format.

    Returns: {sku: {"making_cost": float, "packaging_cost": float}}

    Accepts either:
      - v1: {"SKU": 60}
      - v2: {"SKU": {"making_cost": 60, "packaging_cost": 8}}
      - Mixed: some v1, some v2 — each entry normalised independently.
    """
    if not path.exists():
        return {}

    with open(path, "r", encoding="utf-8") as f:
        raw: dict = json.load(f)

    costs: dict[str, dict] = {}
    for sku, val in raw.items():
        if isinstance(val, dict):
            costs[sku] = {
                "making_cost": val.get("making_cost", 0.0),
                "packaging_cost": val.get("packaging_cost", 0.0),
            }
        elif val is not None:
            # v1 flat format: entire cost treated as making_cost
            costs[sku] = {
                "making_cost": float(val),
                "packaging_cost": 0.0,
            }
        else:
            # null cost — SKU exists in template but cost not filled in
            costs[sku] = None
    return costs


def save_costs(path: Path, costs: dict[str, dict]) -> None:
    """Write costs in v2 format."""
    # Filter out None entries (template placeholders)
    out = {}
    for sku, val in sorted(costs.items()):
        if val is None:
            out[sku] = None
        else:
            out[sku] = {
                "making_cost": val["making_cost"],
                "packaging_cost": val["packaging_cost"],
            }

    path.parent.mkdir(parents=True, exist_ok=True)
    with open(path, "w", encoding="utf-8") as f:
        json.dump(out, f, indent=2, ensure_ascii=False)


def update_single_sku(
    path: Path, sku: str, making_cost: float, packaging_cost: float
) -> dict[str, dict]:
    """Update one SKU's cost and persist. Returns the full updated dict."""
    costs = load_costs(path)
    costs[sku] = {"making_cost": making_cost, "packaging_cost": packaging_cost}
    save_costs(path, costs)
    return costs


def generate_template(skus: set[str]) -> dict:
    """Generate a cost-file template with all given SKUs set to null."""
    return {sku: None for sku in sorted(skus)}


def get_unmapped_skus(
    all_skus: set[str], costs: dict[str, dict]
) -> set[str]:
    """Return SKUs that have no cost entry (missing or null)."""
    return {s for s in all_skus if s not in costs or costs.get(s) is None}

File: frontend/src/services/skuCosts.js
// frontend/src/services/skuCosts.js

const SKU_COSTS_KEY = "meesho_sku_costs";

/**
 * Read the SKU cost JSON from localStorage and normalise every entry to v2 format.
 * Returns: {sku: {"making_cost": float, "packaging_cost": float}}
 */
export function loadCosts() {
  const rawData = localStorage.getItem(SKU_COSTS_KEY);
  if (!rawData) {
    return {};
  }

  try {
    const raw = JSON.parse(rawData);
    const costs = {};

    for (const [sku, val] of Object.entries(raw)) {
      if (val !== null && typeof val === "object") {
        costs[sku] = {
          making_cost: Number(val.making_cost || 0),
          packaging_cost: Number(val.packaging_cost || 0),
        };
      } else if (val !== null) {
        // v1 flat format: entire cost treated as making_cost
        costs[sku] = {
          making_cost: Number(val),
          packaging_cost: 0.0,
        };
      } else {
        // null cost — SKU exists in template but cost not filled in
        costs[sku] = null;
      }
    }
    return costs;
  } catch (err) {
    console.error("Error parsing SKU costs from localStorage", err);
    return {};
  }
}

/**
 * Write costs in v2 format to localStorage.
 */
export function saveCosts(costs) {
  const out = {};
  
  // Sort keys for predictable output
  const sortedSkus = Object.keys(costs).sort();
  
  for (const sku of sortedSkus) {
    const val = costs[sku];
    if (val === null) {
      out[sku] = null;
    } else {
      out[sku] = {
        making_cost: Number(val.making_cost || 0),
        packaging_cost: Number(val.packaging_cost || 0),
      };
    }
  }

  localStorage.setItem(SKU_COSTS_KEY, JSON.stringify(out));
}

/**
 * Update one SKU's cost and persist. Returns the full updated dict.
 */
export function updateSingleSku(sku, makingCost, packagingCost) {
  const costs = loadCosts();
  costs[sku] = {
    making_cost: Number(makingCost || 0),
    packaging_cost: Number(packagingCost || 0),
  };
  saveCosts(costs);
  return costs;
}

/**
 * Generate a cost-file template with all given SKUs set to null.
 */
export function generateTemplate(skus) {
  const template = {};
  Array.from(skus).sort().forEach(sku => {
    template[sku] = null;
  });
  return template;
}

/**
 * Return SKUs that have no cost entry (missing or null).
 */
export function getUnmappedSkus(allSkus, costs) {
  const unmapped = new Set();
  for (const sku of allSkus) {
    if (!(sku in costs) || costs[sku] === null) {
      unmapped.add(sku);
    }
  }
  return unmapped;
}
